File: csharp/Production/utils/Rotation.cs
using CSharpRubikSolver;
using cube;
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.Threading.Tasks;

namespace utils
{
    public class Rotation {

        private Face c_face;
        private Direction c_direction;
        public Rotation() {
        }


        public Rotation(Face p_face, Direction p_direction) {
            c_face = p_face;
            c_direction = p_direction;
        }



        public Rotation(int p_value) {
            if (p_value > 5) {
                c_direction = DirectionFactory.getDirectionByInt(1);
                c_face = FaceHandler.getFace(p_value - 6);
            } else {
                c_direction = DirectionFactory.getDirectionByInt(0);
                c_face = FaceHandler.getFace(p_value);
            }
        }

        public void writeToFile(RubikFileWriter p_write) {
            String l_toWrite = String.Format(" (%d,%d)", (int)c_face, (int)c_direction);
            p_write.write(l_toWrite);
        }

        public Boolean readFromFile(RubikFileReader p_reader) {
            int l_int;
            l_int = p_reader.read();
            //Console.Write("%d ",l_int);
            while ((l_int == ' ') || (l_int == 13 /*'\r'*/)) {
                l_int = p_reader.read();
                // Console.Write("%d ",l_int);
            }

            if ((l_int == /*'\n'*/10) || (l_int == -1 /*EOF*/) || (l_int != '('))
                return false;
            else {

                c_face = FaceHandler.getFace((int)Char.GetNumericValue((char)(p_reader.read())));
                p_reader.read();
                c_direction = DirectionHandler.getDirection((int)Char.GetNumericValue((char)(p_reader.read())));
                p_reader.read();
                return true;
            }
        }


        public int getValue() {
            return ((int)c_face + (int)c_direction * 6);
        }

        public Face getFace() {
            return c_face;
        }

        public Direction getDirection() {
            return c_direction;
        }

        public void print() {
            Console.Write("(%c,%s)", FaceHandler.getCharValue(c_face),DirectionHandler.getString(c_direction));
        }

        public Rotation getReverse() {
            return new Rotation(c_face,DirectionHandler.getOpposite(c_direction));
        }

        public Boolean equals(
                Rotation p_rotation)

        {
            return ((c_face == p_rotation.c_face) &&
                    (c_direction == p_rotation.c_direction));
        }

    public Boolean equals(Object obj) { 
            if (obj == null) {
                return false;
            }
            if (!(new Rotation()).GetType().IsAssignableFrom(obj.GetType())) {
            return false;
        }
        Rotation other = (Rotation) obj;
        if (this.getFace() != other.getFace()) {
            return false;
        }
        if (this.getDirection() != other.getDirection()) {
            return false;
        }
        return true;
        }
    }
}

File: python/production/utils/rotation.py
from production.utils.face import Face
from production.utils.direction import Direction
from production.utils.face_handler import Face_handler
from production.utils.direction_handler import Direction_handler
class Rotation :
        def __init__(self,p_face=None,p_direction=None):
            self.c_face = p_face
            self.c_direction = p_direction

       # def Rotation(self, p_value):
       #     if p_value > 5:
       #         self.c_direction = Direction_factory.getDirectionByInt(1)
       #         self.c_face = Face_handler.getFace(p_value - 6)
       #     else:
       #         self.c_direction = Direction_factory.getDirectionByInt(0)
       #         self.c_face = Face_handler.getFace(p_value)

        def write_to_file(self, p_write):
            l_toWrite = "("+ self.c_face + "," + self.c_direction+")"
            p_write.write(l_toWrite)
        

        def readFromFile(self, p_reader):
            l_int=0
            l_int= p_reader.read()
            #Console.Write("%d ",l_int)
            while (l_int== ' ') or (l_int== 13):
                l_int= p_reader.read()
                # Console.Write("%d ",l_int)
            

            if ((l_int== 10) or (l_int== -1 ) or (l_int!= '(')):
                return False
            else:
                self.c_face = Face_handler.getFaceInt(int(p_reader.read()))
                p_reader.read()
                self.c_direction = Direction_handler.getDirectionInt(int(p_reader.read()))
                p_reader.read()
                return True

        def getValue(self):
            return (self.c_face + self.c_direction * 6)

        def getFace(self):
            return self.c_face

        def getDirection(self):
            return self.c_direction

        def print(self):
            print("(" + Face_handler.getCharValue(self.c_face) + "," + Direction_handler.getString(self.c_direction) + "\n")

        def getReverse(self):
            return Rotation(self.c_face, Direction_handler.getOpposite(self.c_direction))

        def equals(self, p_rotation):
            return ((self.c_face == p_rotation.getFace()) and
                    (self.c_direction == p_rotation.getDirection()))
        

       
        
    

